File: go/puzzle.go
package main

import (
	"fmt"
	pq "github.com/emirpasic/gods/queues/priorityqueue"
	"github.com/emirpasic/gods/utils"
	"log"
	"os"
	"runtime"
	"slices"
	"strconv"
	"strings"
	"sync"
	"time"
)

type Action int

const (
	_ Action = iota
	Up
	Down
	Left
	Right
)

func (action Action) String() string {
	switch action {
	case Down:
		return "Down\n"
	case Up:
		return "Up\n"
	case Left:
		return "Left\n"
	case Right:
		return "Right\n"
	default:
		return "Start\n"
	}
}

type Position struct {
	row int
	col int
}

func IndexToPos(i int, dimension int) Position {
	return Position{row: i / dimension, col: i % dimension}
}

func (pos Position) ToIndex(dimension int) int {
	return pos.row*dimension + pos.col
}

func (pos Position) InBounds(dimension int) bool {
	return pos.row >= 0 && pos.row < dimension && pos.col >= 0 && pos.col < dimension
}

func (pos Position) Add(rhs Position) Position {
	return Position{row: pos.row + rhs.row, col: pos.col + rhs.col}
}

type Tile = uint8

type Puzzle struct {
	prev      *Puzzle
	tiles     []Tile
	g         int
	f         int
	action    Action
	dimension int
}

func NewGoal(len int) []Tile {
	tiles := make([]Tile, len)
	for i := range len {
		tiles[i] = byte(i)
	}
	return tiles
}

func (puzzle *Puzzle) PrintPuzzle() string {
	var sb strings.Builder
	for i, tile := range puzzle.tiles {
		if tile == 0 {
			sb.WriteString("  ")
		} else {
			fmt.Fprintf(&sb, "%d ", tile)
		}
		if (i+1)%puzzle.dimension == 0 {
			sb.WriteString("\n")
		}
	}
	return sb.String()
}

func abs(i int) int {
	if i < 0 {
		return -i
	}
	return i
}

func (puzzle *Puzzle) Heuristic() int {
	h := 0
	for i, tile := range puzzle.tiles {
		if tile == 0 {
			continue
		}
		pos1 := IndexToPos(i, puzzle.dimension)
		pos2 := IndexToPos(int(tile), puzzle.dimension)
		h += abs(pos2.row-pos1.row) + abs(pos2.col-pos1.col)
	}
	return h
}

func (puzzle *Puzzle) FindZero() Position {
	for i, tile := range puzzle.tiles {
		if tile == 0 {
			return IndexToPos(i, puzzle.dimension)
		}
	}
	panic("Puzzle contains no zero - this should never happen")
}

func HashTiles(tiles []Tile) uint64 {
	var hash uint64
	for i, tile := range tiles {
		mask := (uint64(tile)) << (i * 4)
		hash = hash | mask
	}
	return hash
}

func (puzzle *Puzzle) Hash() uint64 {
	return HashTiles(puzzle.tiles)
}

type Direction struct {
	pos    Position
	action Action
}

var directions = []Direction{
	{pos: Position{row: 0, col: 1}, action: Right},
	{pos: Position{row: 0, col: -1}, action: Left},
	{pos: Position{row: 1, col: 0}, action: Down},
	{pos: Position{row: -1, col: 0}, action: Up},
}

func (puzzle *Puzzle) OnNeighbors(onNeighbor func(puzzle Puzzle)) {
	zeroPos := puzzle.FindZero()
	for _, direction := range directions {
		nextPos := zeroPos.Add(direction.pos)
		if !nextPos.InBounds(puzzle.dimension) {
			continue
		}

		nextPuzzle := Puzzle{
			prev:      puzzle,
			tiles:     slices.Clone(puzzle.tiles),
			g:         puzzle.g + 1,
			dimension: puzzle.dimension,
			action:    direction.action,
		}

		zeroIdx := zeroPos.ToIndex(puzzle.dimension)
		nextIdx := nextPos.ToIndex(puzzle.dimension)

		temp := nextPuzzle.tiles[nextIdx]
		nextPuzzle.tiles[nextIdx] = nextPuzzle.tiles[zeroIdx]
		nextPuzzle.tiles[zeroIdx] = temp

		onNeighbor(nextPuzzle)
	}
}

func ReconstructPath(puzzle *Puzzle) []Puzzle {
	path := make([]Puzzle, 0)
	for puzzle != nil {
		path = append(path, *puzzle)
		puzzle = puzzle.prev
	}
	slices.Reverse(path)
	return path
}

func FindPath(initial Puzzle) ([]Puzzle, int) {
	visited := make(map[uint64]struct{})

	frontier := pq.NewWith(func(a, b interface{}) int {
		priorityA := a.(*Puzzle).f
		priorityB := b.(*Puzzle).f
		return utils.IntComparator(priorityA, priorityB)
	})

	frontier.Enqueue(&initial)

	goal := NewGoal(len(initial.tiles))
	goalHash := HashTiles(goal)

	nodes := 0
	for {
		top, ok := frontier.Dequeue()
		if !ok {
			break
		}
		puzzle := top.(*Puzzle)
		nodes += 1

		currHash := puzzle.Hash()
		visited[currHash] = struct{}{}

		if currHash == goalHash {
			return ReconstructPath(puzzle), nodes
		}

		puzzle.OnNeighbors(func(puzzle Puzzle) {
			_, isVisited := visited[puzzle.Hash()]
			if !isVisited {
				puzzle.f = puzzle.g + puzzle.Heuristic()
				frontier.Enqueue(&puzzle)
			}
		})
	}

	return make([]Puzzle, 0), nodes
}

func ReadPuzzles(path string) ([]Puzzle, error) {
	contents, err := os.ReadFile(path)
	if err != nil {
		return nil, fmt.Errorf("failed to read input file: %v\n", err)
	}

	var puzzles []Puzzle
	var current []Tile

	for line := range strings.SplitSeq(string(contents), "\n") {
		line := strings.TrimRight(line, "\n\r")
		tokens := strings.Split(line, " ")

		if len(tokens) > 1 {
			for _, token := range tokens {
				if token == "" {
					continue
				}
				tile, err := strconv.Atoi(token)
				if err != nil {
					return nil, fmt.Errorf("tile %s must be integers", token)
				}
				current = append(current, byte(tile))
			}
		} else {
			if len(current) == 0 {
				continue
			}

			var dimension int
			switch len(current) {
			case 9:
				dimension = 3
			case 16:
				dimension = 4
			default:
				return nil, fmt.Errorf("puzzle must have 9 or 16 tiles, got %d", len(current))
			}

			puzzles = append(puzzles, Puzzle{prev: nil, tiles: current, dimension: dimension})
			current = make([]Tile, 0)
		}
	}

	return puzzles, nil
}

type Solution struct {
	time  float64
	nodes int
	path  []Puzzle
}

func FindPaths(puzzles []Puzzle) []Solution {
	var solutions []Solution
	for _, puzzle := range puzzles {
		start := time.Now()

		solution, nodes := FindPath(puzzle)

		duration := time.Since(start).Microseconds()
		t := float64(duration) / 1000.0

		solutions = append(solutions, Solution{
			time:  t,
			nodes: nodes,
			path:  solution,
		})
	}
	return solutions
}

func FindPathsParallel(puzzles []Puzzle) []Solution {
	solutions := make([]Solution, len(puzzles))

	var wg sync.WaitGroup
	sem := make(chan struct{}, runtime.NumCPU())

	for i, puzzle := range puzzles {
		wg.Go(func() {
			sem <- struct{}{}

			start := time.Now()
			solution, nodes := FindPath(puzzle)

			duration := time.Since(start).Microseconds()
			t := float64(duration) / 1000.0

			solutions[i] = Solution{
				time:  t,
				nodes: nodes,
				path:  solution,
			}

			<-sem
		})
	}

	wg.Wait()
	return solutions
}

func main() {
	if len(os.Args) < 2 {
		log.Fatalf("Needs at least argument for input file")
	}

	inputFile := os.Args[1]

	flag := "seq"
	if len(os.Args) >= 3 {
		flag = os.Args[2]
	}

	puzzles, err := ReadPuzzles(inputFile)
	if err != nil {
		log.Fatalf("failed to read puzzles: %s", err)
	}

	start := time.Now()

	var solutions []Solution
	switch flag {
	case "seq":
		solutions = FindPaths(puzzles)
	case "par":
		solutions = FindPathsParallel(puzzles)
	default:
		log.Fatalf("Parallelism flag must be par or seq, got %s \n", flag)
	}

	duration := time.Since(start).Microseconds()
	eteTime := float64(duration) / 1000.0

	for i, s := range solutions {
		fmt.Printf("Solution for puzzle %d\n", i+1)
		for _, puzzle := range s.path {
			fmt.Print(puzzle.action.String())
		}

		fmt.Printf("Solved in %d steps\n\n", len(s.path)-1)
	}

	var totalTime float64
	var totalNodes int
	for i, s := range solutions {
		fmt.Printf("Puzzle %d: %f ms, %d nodes\n", i+1, s.time, s.nodes)
		totalTime += s.time
		totalNodes += s.nodes
	}
	fmt.Printf("\nTotal: %f ms, %d nodes\n", totalTime, totalNodes)

	fmt.Printf("End-to-end: %f ms\n", eteTime)
}

File: nim/puzzle.nim
import std/heapqueue
import std/tables
import std/strutils
import std/bitops
import std/times
import std/os
import std/strformat
import std/algorithm

func puzzleLen(D: int): int = D * D - 1

type 
    Action = enum Start Up Down Left Right
    Position = object
        row: int
        col: int
    Direction = object
        position: Position
        action: Action
    Dimension = static[int]
    Tiles[D: Dimension] = array[0..puzzleLen(D), 0..puzzleLen(D)]
    Puzzle[D: Dimension] {.byref.} = object
        prev: ref Puzzle[D]
        tiles: Tiles[D]
        action: Action
        f: int
        g: int  
    EightPuzzle = Puzzle[3]
    FifteenPuzzle = Puzzle[4]
    Solution[D: Dimension] = object
        time: float
        nodes: int
        path: seq[Puzzle[D]]
    PuzzleKind = enum EightPuzzleKind FifteenPuzzleKind
    AnyPuzzle = object
        case kind: PuzzleKind
        of EightPuzzleKind:
            eight: EightPuzzle
        of FifteenPuzzleKind:
            fifteen: FifteenPuzzle
    AnySolution = object
        duration: Duration
        nodes: int
        case kind: PuzzleKind
        of EightPuzzleKind:
            eightPath: seq[EightPuzzle]
        of FifteenPuzzleKind:
            fifteenPath: seq[FifteenPuzzle]
    
proc puzzleFromTiles[D: Dimension](tilesSeq: seq[int]): Puzzle[D] =
    var tilesArr: Tiles[D]
    for i, tile in tilesSeq.pairs:
        # throws `RangeDefect` if the tile is not in the tilesArr ordinal range
        tilesArr[i] = tile
    return Puzzle[D](tiles: tilesArr)

proc newPuzzle[D: Dimension](puzzle: Puzzle[D]): ref Puzzle[D] =
    var puzzleRef = new Puzzle[D]
    puzzleRef[] = puzzle
    return puzzleRef

func indexToPos[D: Dimension](index: int): Position = Position(row: int(index / D), col: index mod D)

func posToIndex[D: Dimension](pos: Position): int = pos.row * D + pos.col

func isPositionInBounds[D: Dimension](pos: Position): bool = pos.row >= 0 and pos.row < D and pos.col >= 0 and pos.col < D

proc add (a, b: Position): Position = Position(row: a.row + b.row, col: a.col + b.col)  

proc puzzleToString[D: Dimension](puzzle: Puzzle[D]): string = 
    var str = puzzle.action.actionToString()
    str.add "\n"

    for i, tile in puzzle.tiles.pairs:
        if tile == 0:
            str.add "  "
        else:
            str.add $tile
            str.add " "
        if (i+1) mod D == 0:
            str.add "\n"

    return str

proc puzzleSeqToString[D: Dimension](puzzles: seq[Puzzle[D]]): string = 
    var str = ""
    for puzzle in puzzles:
        str.add &"{puzzle}\n"
    return str

func actionToString(action: Action): string = 
    case action:
        of Start:
            result = "Start"
        of Up:
            result = "Up"
        of Down:
            result = "Down"
        of Right:
            result = "Right"
        of Left:
            result = "Left"

proc printPath[D: Dimension](path: seq[Puzzle[D]]): string = 
    var str = ""
    for i, puzzle in path.pairs:
        str.add (actionToString puzzle.action)
        if i != path.len()-1:
            str.add "\n"
    return str

proc printSolution(solution: AnySolution) =
    case solution.kind:
        of EightPuzzleKind:
            echo printPath(solution.eightPath)
        of FifteenPuzzleKind:
            echo printPath(solution.fifteenPath)

proc steps(solution: AnySolution): int = 
    case solution.kind:
        of EightPuzzleKind:
            return solution.eightPath.len()-1
        of FifteenPuzzleKind:
            return solution.fifteenPath.len()-1
                
proc newGoal[D: Dimension](): Tiles[D] = 
    var tiles: Tiles[D]
    for i in 0..puzzleLen(D):
        tiles[i] = i
    return tiles

proc hashTiles[D: Dimension](tiles: Tiles[D]): uint64 = 
    var hash: uint64

    for i, tile in tiles.pairs:
        let mask = uint64(tile) shl (i * 4)
        hash = bitor(hash, mask)

    return hash

proc hashPuzzle[D: Dimension](puzzle: Puzzle[D]): uint64 = 
    return hashTiles(puzzle.tiles)

proc heuristic[D: Dimension](puzzle: Puzzle[D]): int =
    var h = 0

    for i, tile in puzzle.tiles.pairs:
        if tile == 0:
            continue
        let pos1 = indexToPos[D](i)
        let pos2 = indexToPos[D](tile)
        h += abs(pos2.row - pos1.row) + abs(pos2.col - pos1.col)

    return h 

proc findZero[D: Dimension](puzzle: Puzzle[D]): Position =
    for i, tile in puzzle.tiles.pairs:
        if tile == 0:
            return indexToPos[D](i)
    raise newException(AssertionDefect, "A puzzle must contain at least one zero")

proc reconstructPath[D: Dimension](puzzle: ref Puzzle[D]): seq[Puzzle[D]] =
    var path: seq[Puzzle[D]]
    var currPuzzle = puzzle

    while not isNil currPuzzle:
        path.add currPuzzle[]
        currPuzzle = currPuzzle.prev
    
    path.reverse()
    return path

var directions = [
    Direction(position: Position(row: -1, col: 0), action: Up),
    Direction(position: Position(row: 1, col: 0), action: Down),
    Direction(position: Position(row: 0, col: 1), action: Right),
    Direction(position: Position(row: 0, col: -1), action: Left)
] 

iterator neighbors[D: Dimension](puzzle: ref Puzzle[D]): Puzzle[D] = 
    let zeroPos = (puzzle[]).findZero()
    for direction in directions:
        let nextPos = zeroPos.add direction.position
        if not isPositionInBounds[D](nextPos):
            continue
        
        var nextPuzzle = Puzzle[D](prev: puzzle, g: puzzle.g + 1, action: direction.action)
        nextPuzzle.tiles = puzzle.tiles

        let zeroIdx = posToIndex[D](zeroPos)
        let nextIdx = posToIndex[D](nextPos)

        swap(nextPuzzle.tiles[zeroIdx], nextPuzzle.tiles[nextIdx])

        yield nextPuzzle

proc `<` [D: Dimension](a, b: ref Puzzle[D]): bool = a.f < b.f

proc findPath[D: Dimension](inputPuzzle: Puzzle[D]): Solution[D] =
    var goal = newGoal[D]()
    let goalHash = hashTiles goal

    var visited = initTable[uint64, bool]()

    let rootPuzzle = newPuzzle inputPuzzle

    var frontier = initHeapQueue[ref Puzzle[D]]()
    frontier.push rootPuzzle

    var nodes = 0

    while frontier.len > 0:
        var items: seq[Puzzle[D]] = @[]
        for item in frontier.items():
            items.add item[]

        let currPuzzle = frontier.pop()
        nodes += 1
      
        let currHash = hashPuzzle currPuzzle[]
        visited[currHash] = true    

        if currHash == goalHash:
            return Solution[D](path: reconstructPath currPuzzle, nodes: nodes)

        for p in currPuzzle.neighbors():
            var puzzle = p
            let isVisited = visited.hasKey (hashPuzzle puzzle)
            if not isVisited:
                puzzle.f = puzzle.g + puzzle.heuristic()
                frontier.push (newPuzzle puzzle)

    return Solution[D](path: @[], nodes: nodes)

proc findPaths(puzzles: seq[AnyPuzzle]): seq[AnySolution] =
    var solutions: seq[AnySolution] = @[]

    for puzzle in puzzles:
        var anySolution: AnySolution

        let startTime = now()
        case puzzle.kind:
            of EightPuzzleKind:
                let solution = findPath puzzle.eight
                anySolution = AnySolution(kind: EightPuzzleKind, eightPath: solution.path, nodes: solution.nodes)
            of FifteenPuzzleKind:
                let solution = findPath puzzle.fifteen
                anySolution = AnySolution(kind: FifteenPuzzleKind, fifteenPath: solution.path, nodes: solution.nodes)

        anySolution.duration = now() - startTime

        solutions.add anySolution

    return solutions

proc readPuzzles(path: string): seq[AnyPuzzle] =
    let contents = readFile(path)

    var tiles: seq[int] = @[]
    var puzzles: seq[AnyPuzzle] = @[]

    for lineItr in contents.split("\n"):
        let line = lineItr.strip()
        let tokens = line.split(" ")
        if tokens.len() > 1:
            for token in tokens:
                if token == "":
                    continue
                tiles.add (parseInt token)
        else:
            var puzzle: AnyPuzzle
            case tiles.len():
                of 0:
                    continue
                of 9:
                    puzzle = AnyPuzzle(kind: EightPuzzleKind, eight: puzzleFromTiles[3](tiles))
                of 16:
                    puzzle = AnyPuzzle(kind: FifteenPuzzleKind, fifteen: puzzleFromTiles[4](tiles))
                else:
                    raise newException(ValueError, "A puzzle must be of length 9 or 16")
            puzzles.add puzzle
            tiles.setLen(0)

    return puzzles

func microseconds(duration: Duration): float = float(duration.inMicroseconds()) / 1000.0 

let params = commandLineParams()
if params.len() < 2:
    echo "Expected at least one argument for input file"
    quit(1)

let inputFilePath = params[1]

let inputPuzzles = readPuzzles(inputFilePath)

let startTime = now()

let solutions = findPaths(inputPuzzles)

let e2eTime = now() - startTime

for i, solution in solutions.pairs:
    echo &"Solution for puzzle {i+1}"
    printSolution(solution)
    echo &"Solved in {solution.steps()} steps"

var
    totalTime: Duration
    totalNodes: int

echo ""
for i, solution in solutions.pairs:
    echo &"Puzzle {i+1}: {microseconds(solution.duration)} ms, {solution.nodes} nodes"
    totalTime += solution.duration
    totalNodes += solution.nodes

echo &"\nTotal: {microseconds(totalTime)} ms, {totalNodes} nodes"
echo &"End-to-end: {microseconds(e2eTime)} ms"